crates/ruff_linter/src/checkers/ast/mod.rs

```diff
@@ -1461,6 +1461,11 @@ impl<'a> Visitor<'a> for Checker<'a> {
             _ => {}
         };
 
+        // Restore boolean test flag after examining subexpressions
+        if flags_snapshot.intersects(SemanticModelFlags::BOOLEAN_TEST) {
+            self.semantic.flags |= SemanticModelFlags::BOOLEAN_TEST;
+        }
+
         // Step 4: Analysis
         analyze::expression(expr, self);
         match expr {
```

Comment:
Thanks for the explanation. The change now makes sense to me. However, I do think that this problem isn't specific to `BOOLEAN_TEST` but applies to all flags. 
The flags should be restored when visiting the expression itself. They're only relevant when visiting the sub-expressions. 

That's why I suggest that we change the code to:

```rust
        self.semantic.flags = flags_snapshot;
        // Step 4: Analysis
        match expr {
            Expr::StringLiteral(string_literal) => {
                analyze::string_like(string_literal.into(), self);
            }
            Expr::BytesLiteral(bytes_literal) => analyze::string_like(bytes_literal.into(), self),
            Expr::FString(f_string) => analyze::string_like(f_string.into(), self),
            _ => {}
        }
```

Reply:
I skimmed through the `SemanticFlags` and I didn't spot any expression level flag that would change the flags for the parent or a sibling.